apps/web/src/components/Table/ResponsiveTable/ResponsiveTable.tsx

```diff
@@ -0,0 +1,335 @@
+  /** Enable sorting */
+  sort?: SortDef;
+  /** Enable pagination */
+  pagination?: PaginationDef;
+  /** Enable printing selected rows (requires rowSelect) */
+  print?: DatasetPrint;
+  /** Enable downloading selected rows and/or all data (requires rowSelect) */
+  download?: DatasetDownload;
+  /** Enable the "add item" button */
+  add?: AddLinkProps;
+  /** Filter component */
+  filterComponent?: React.ReactNode;
+}
+
+const ResponsiveTable = <TData extends object>({
+  caption,
+  data,
+  columns,
+  hiddenColumnIds,
+  isLoading,
+  nullMessage,
+  rowSelect,
+  search,
+  sort,
+  download,
+  print,
+  add,
+  pagination,
+  filterComponent,
+}: TableProps<TData>) => {
+  const id = React.useId();
+  const intl = useIntl();
+  const [, setSearchParams] = useSearchParams();
+  const isInternalSearch = search && search.internal;
+  const memoizedColumns = React.useMemo(() => {
+    if (!rowSelect) return columns;
+    // Inject the selection column if it is enabled
+    return [getRowSelectionColumn(rowSelect.cell, intl), ...columns];
+  }, [columns, intl, rowSelect]);
+  const columnIds = memoizedColumns.map((column) => column.id).filter(notEmpty);
+
+  const [rowSelection, setRowSelection] = useRowSelection<TData>(
+    data,
+    rowSelect,
+  );
+  const { state, initialState, updaters } = useControlledTableState({
+    columnIds,
+    initialState: {
+      hiddenColumnIds,
+      searchState: search?.initialState,
+      sortState: sort?.initialState,
+      paginationState: pagination?.initialState,
+    },
+  });
+
+  const manualPageSize = !pagination?.internal
+    ? Math.ceil(
+        (pagination?.total ?? 0) /
+          (state.pagination?.pageSize ??
+            INITIAL_STATE.paginationState.pageSize),
+      )
+    : undefined;
+
+  const table = useReactTable({
+    data,
+    columns: memoizedColumns,
+    state: {
+      ...state,
+      rowSelection,
+    },
+    enableGlobalFilter: isInternalSearch,
+    enableRowSelection: !!rowSelect,
+    enableSorting: !!sort,
+    manualSorting: !sort?.internal,
+    manualPagination: !pagination?.internal,
+    pageCount: manualPageSize,
+    getCoreRowModel: getCoreRowModel(),
+    getFilteredRowModel: getFilteredRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    getPaginationRowModel: getPaginationRowModel(),
+    onRowSelectionChange: setRowSelection, // Note: We should probably do the state sync here
+    ...updaters,
+  });
+
+  const searchColumns = table
+    .getAllLeafColumns()
+    .filter((column) => column.getCanFilter())
+    .map((column) => ({
+      label: getColumnHeader(column, "searchHeader"),
+      value: column.id,
+    }));
+
+  const sortingState = table.getState().sorting;
+  const columnFilterState = table.getState().columnFilters;
+  const globalFilterState = table.getState().globalFilter;
+  const columnVisibilityState = table.getState().columnVisibility;
+  const paginationState = table.getState().pagination;
```

Comment:
I agree! I think this was a similar oddity I was doing as a way to prevent infinite re-renders. Seems like I was dead wrong.

Updated in [update state to use deconstruction](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7588/commits/ee07e87336ddab0aa3df8b407671dec041fa337f)